Don't use pregenerated files anymore

fingolfin opened the pull request:
... thanks to the [recent cross compilation fixes in Singular master](https://github.com/Singular/Sources/commit/26b42ddcca6857a388490aa8398e5ed81434ab0e).

In principle we could now build for more target platforms, however
not all our dependencies (Flint...) provide for more platforms, so
we can't.

@hannes14 @wbhart how important is the NTL dependency? are there things it provides that are not (yet) covered by FLINT?

fingolfin commented:
This doesn't quite work, as the changes in spielwiese are incomplete; we also need https://github.com/Singular/Sources/pull/1003

fingolfin commented:
I temporarily modified this to grab a revision from my repo containing https://github.com/Singular/Sources/pull/1003, just to see if it works with that.

wbhart commented:
What platforms do you want to build for that Flint doesn't support!? 

NTL's factorisation of poly's over Z is still faster than Flint for small numbers of local factors. Our Zassenhaus has been improved so it's not as bad as it was, so perhaps some benchmarking is in order. Perhaps we can live with it. No one is going to fund development of a fast pruned Zassenhaus implementation a la NTL so it is probably now a permanent situation.

Of course NTL is also a small constant factor faster than Flint on some benchmarks and vice versa.

I'm not sure I can think of anything that NTL does that Flint still doesn't. We even have root finding now. Actually I vaguely recall NTL does BKZ reduction and Flint does not. But Singular probably doesn't use that.

fingolfin commented:
The problem is not what Flint supports, but rather what https://github.com/thofma/Flint2Builder provides, and that's not much right now (although I am sure Tommy could add more -- but it'd be easier to just switch to https://github.com/JuliaBinaryWrappers/FLINT_jll.jl)

However, NTL is much more of a problem, as it is rather difficult to build with a cross compiler (basically one has to replace a noticeable chunk of the build system). So I wonder if we can just drop it, and use Flint instead. But for that, it's not just a question of what Flint can do; but also one of whether Singular actually use it. I mean, even if feature X is present in both NTL and Flint, it might be that Singular only uses it from NTL (simply because nobody so far added code for calling the relevant Flint APIs).

fingolfin commented:
Note to self: the build issue is due to -Qunused-arguments ... resolve that tomorrow

thofma commented:
Indeed, it is not a question of available functionality that could be used sometime in the future. It is more the question of the current Singular version being useful without NTL?

hannes14 commented:
Beside some experimental stuff (we can live without it: HNF) NTL is used
for Z-matrices: LLL
for univariate polynomials::
 - absolute factorization (factoring over Q(alpha) and Z/p(alpa))
 - factorization over Q and Z/p (partially alternative exists)
 - gcd stuff: ezgcd, BuildIrred, ...
 - multiplication over Z/p and Z/p(alpa) (if FLINT is present, FLINT is used)
 - division with/without rest over Z , Z/p, Z/p(alpha) (if FLINT is present, FLINT is used)

fingolfin commented:
@hannes14 so, do I understand it right, you are saying that Singular uses NTL (if available) for the following tasks, but not Flint?
- LLL for Z matrices
- absolute factorization (factoring over Q(alpha) and Z/p(alpa)) for univariate polynomials
- factorization over Q and Z/p for univariate polynomials (except in some cases?!? what are the "partially existing alternatives"?)
- gcd stuff: ezgcd, BuildIrred, ...

If so, is this because Flint doesn't provide these (or perhaps didn't provide them before 2.6)? Or because nobody wrote the relevant code yet? @wbhart ?

I think it would be really nice to be able to use Flint as a full replacement for NTL (ignoring the caveats Bill already explained). I don't mind having code which optionally uses NTL, but within the cornerstone, it seems hard to justify why we need to rely on NTL when we have Flint.

wbhart commented:
> Beside some experimental stuff (we can live without it: HNF) NTL is used
> for Z-matrices: LLL

Flint has HNF and LLL. Both should be comparable to NTL.

> for univariate polynomials::
> 
> * absolute factorization (factoring over Q(alpha) and Z/p(alpa))

Flint does not have absolute factorisation. I was unaware NTL had it either.

> * factorization over Q and Z/p (partially alternative exists)

Flint has this.

> * gcd stuff: ezgcd, BuildIrred, ...

I don't know what BuildIrred is. We certainly have GCD, both univariate and multivariate, but I was unaware NTL had ezgcd. Our GCD will be uniformly better.

> * multiplication over Z/p and Z/p(alpa) (if FLINT is present, FLINT is used)
> * division with/without rest over Z , Z/p, Z/p(alpha) (if FLINT is present, FLINT is used)

I guess you mean remainder?

hannes14 commented:
- factorization over Z/p: code to use NTL or FLINT exists in factory
-                      over Q (Z):  NTL 
-                      over  Z/p(alpha), p>2: NTL or FLINT, p==2: NTL
-                      over Q(alpha): factory

fingolfin commented:
This could be merged now. Perhaps someone could transfer the discussion (resp. a summary of it) about replacing NTL usage with Flint in Singular, to the Singular issue tracker?

fingolfin committed:
Don't use pregenerated files anymore

... thanks to the recent cross compilation fixes in Singular master.

In principle we could now build for more target platforms, however
not all our dependencies (Flint...) provide for more platforms, so
we can't.
